pkg/odo/cli/run/run.go

```diff
@@ -0,0 +1,119 @@
+package run
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/devfile/library/v2/pkg/devfile/parser/data/v2/common"
+	"github.com/spf13/cobra"
+
+	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/odo/cli/errors"
+	"github.com/redhat-developer/odo/pkg/odo/cmdline"
+	"github.com/redhat-developer/odo/pkg/odo/commonflags"
+	fcontext "github.com/redhat-developer/odo/pkg/odo/commonflags/context"
+	odocontext "github.com/redhat-developer/odo/pkg/odo/context"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions/clientset"
+	odoutil "github.com/redhat-developer/odo/pkg/odo/util"
+	"github.com/redhat-developer/odo/pkg/podman"
+	scontext "github.com/redhat-developer/odo/pkg/segment/context"
+
+	ktemplates "k8s.io/kubectl/pkg/util/templates"
+)
+
+const (
+	RecommendedCommandName = "run"
+)
+
+type RunOptions struct {
+	// Clients
+	clientset *clientset.Clientset
+
+	// Args
+	commandName string
+}
+
+var _ genericclioptions.Runnable = (*RunOptions)(nil)
+
+func NewRunOptions() *RunOptions {
+	return &RunOptions{}
+}
+
+var runExample = ktemplates.Examples(`
+	# Run the command "my-command" in the Dev mode
+	%[1]s my-command
+
+`)
+
+func (o *RunOptions) SetClientset(clientset *clientset.Clientset) {
+	o.clientset = clientset
+}
+
+func (o *RunOptions) Complete(ctx context.Context, cmdline cmdline.Cmdline, args []string) error {
+	o.commandName = args[0] // Value at 0 is expected to exist, thanks to ExactArgs(1)
+	return nil
+}
+
+func (o *RunOptions) Validate(ctx context.Context) error {
+	var (
+		devfileObj = odocontext.GetEffectiveDevfileObj(ctx)
+		platform   = fcontext.GetPlatform(ctx, commonflags.PlatformCluster)
+	)
+
+	if devfileObj == nil {
+		return genericclioptions.NewNoDevfileError(odocontext.GetWorkingDirectory(ctx))
+	}
+
+	commands, err := devfileObj.Data.GetCommands(common.DevfileOptions{
+		FilterByName: o.commandName,
+	})
+	if err != nil || len(commands) != 1 {
+		return errors.NewNoCommandNameInDevfileError(o.commandName)
```

Comment:
Can we separate the cases where `err != nil` and `len(commands)!= 1`? In the first case, I think it would make more sense to return `err` as it might provide the user with additional details about the error (other than just a command not found error)..